packages/concerto-core/lib/introspect/mapvaluetype.js

```diff
@@ -0,0 +1,127 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const Decorated = require('./decorated');
+const IllegalModelException = require('./illegalmodelexception');
+
+// Types needed for TypeScript generation.
+/* eslint-disable no-unused-vars */
+/* istanbul ignore next */
+if (global === undefined) {
+    const ModelFile = require('./modelfile');
+}
+
+/**
+ * MapValueType defines a Value type of MapDeclaration.
+ *
+ * @extends Decorated
+ * @see See {@link Decorated}
+ * @class
+ * @memberof module:concerto-core
+ */
+class MapValueType extends Decorated {
+    /**
+     * Create an MapValueType.
+     * @param {MapDeclaration} parent - The owner of this property
+     * @param {Object} ast - The AST created by the parser
+     * @param {ModelFile} modelFile - the ModelFile for the Map class
+     * @throws {IllegalModelException}
+     */
+    constructor(parent, ast) {
+        super(ast);
+        this.parent = parent;
+        this.type = null;
+        this.process();
+    }
+
+    /**
+     * Process the AST and build the model
+     *
+     * @throws {IllegalModelException}
+     * @private
+     */
+    process() {
+        super.process();
+        this.type = this.ast.name;
+    }
+
+    /**
+     * Semantic validation of the structure of this class.
+     *
+     * @throws {IllegalModelException}
+     * @protected
+     */
+    validate() {
+        const declarations = this.getModelFile().getAllDeclarations();
+
+        const value = declarations.find(decl => decl.name === this.type);
+
+        if (!value?.isConcept?.()           &&
+            !value?.isEnum?.()              &&
+            !value?.isAsset?.()             &&
+            !value?.isEvent?.()             &&
+            !value?.isParticipant?.()       &&
+            !value?.isTransaction?.()       &&
+            !value?.isMapDeclaration?.()    &&
+            !value?.isScalarDeclaration?.() &&
+            !['String', 'Long', 'Integer', 'Double', 'Boolean', 'DateTime'].includes(this.type)) {
+            throw new IllegalModelException(
+                `MapPropertyType has invalid Type: ${this.type}`
+            );
+        }
+    }
+
+    /**
+     * Returns the ModelFile that defines this class.
+     *
+     * @public
+     * @return {ModelFile} the owning ModelFile
+     */
+    getModelFile() {
+        return this.parent.getModelFile();
+    }
+
+    /**
+     * Returns the name of the MapValue. This name does not include the
+     * namespace from the owning ModelFile.
+     *
+     * @return {string} the short name of this class
+     */
+    getName() {
+        return this.ast.name;
+    }
+
+    /**
+     * Returns the Type of the MapValue. This name does not include the
+     * namespace from the owning ModelFile.
+     *
+     * @return {string} the short name of this class
+     */
+    getType() {
+        return this.ast.name;
```

Comment:
Should be `this.ast.type`?

Reply:
see above. This is set to return a global variable now, but is assigned from the same property